### PR TITLE
fix issue with templates defined by deps being required while still parsing deps

### PR DESCRIPTION
**Pull request by boegel**
This should fix the problems being reported in #1402 and https://github.com/hpcugent/easybuild-easyconfigs/issues/3210.

cc @verdurin, @ocaisa

### CODE REVIEW

**Comment by boegel** [comment]
@verdurin, @ocaisa: I'll look into an test that covers this, but it would be great if you could verify this fix (and review the changes).

**Comment by verdurin** [comment]
@boegel having manually applied the new version of `easyconfig.py` then reload the module, I tried this:

```
eb cnvkit-0.7.3-foss-2015b-Python-2.7.10.eb -dr --force --minimal-toolchains --robot ./
```

and it worked, without generating the errors I reported in https://github.com/hpcugent/easybuild-easyconfigs/issues/3210

File: easybuild/framework/easyconfig/easyconfig.py [hunk 1]
```diff
@@ -806,19 +811,8 @@ def _parse_dependency(self, dep, hidden=False, build_only=False):
...
-                orig_tc = tc
-                tc = robot_find_minimal_toolchain_of_dependency(dependency, self.modules_tool, parent_tc=tc)
-                if tc is None:
-                    raise EasyBuildError("No easyconfig for %s that matches toolchain hierarchy generated by %s",
-                                         dependency, orig_tc)
-                else:
-                    self.log.debug("Obtained minimal toolchain: %s", tc)
-            else:
-                self.log.debug("Inheriting parent toolchain %s for dependency %s", tc, dependency)
+            self.log.debug("Inheriting parent toolchain %s for dependency %s (for now)", tc, dependency)
```
Comment (verdurin) [inline easybuild/framework/easyconfig/easyconfig.py]: @boegel - can you clarify what you mean by "(for now)" here? This is an interim fix, or just "for now" as part of the dependency resolution process?

File: easybuild/framework/easyconfig/easyconfig.py [hunk 2]
```diff
@@ -843,28 +837,70 @@ def _parse_dependency(self, dep, hidden=False, build_only=False):
...
+
+        for key in DEPENDENCY_PARAMETERS:
+            # loop over a *copy* of dependency dicts (with resolved templates);
+            # to update the original dep dict, we need to index with idx into self._config[key][0]...
+            for idx, dep in enumerate(self[key]):
+
+                # reference to original dep dict, this is the one we should be updating
+                orig_dep = self._config[key][0][idx]
+
+                # minimize toolchains for dependencies with inherite toolchain, if requested
```
Comment (verdurin) [inline easybuild/framework/easyconfig/easyconfig.py]: Should be "inherited"?

**Comment by ocaisa** [comment]
@boegel Tested with `Boost-1.60.0-intel-2016a-Python-2.7.11.eb` from the main easyconfig repo and parses everything correctly.

File: easybuild/framework/easyconfig/easyconfig.py (hunk 1 above)
Comment (ocaisa) [inline easybuild/framework/easyconfig/easyconfig.py]: With `minimal-toolchains` you might be replacing this toolchain with one further down the hierarchy when the dependencies are finalised. Probably could do with replacing `(for now)` by `(until dependencies are finalised)`

**Comment by ocaisa** [comment]
lgtm

**Comment by wpoely86** [comment]
:+1:

**Comment by boegel** [comment]
Thanks for the confirmation of the fix @verdurin and @ocaisa, and for the review of the changes!

Going in!

**Pull request merged**